### DISCUSSION TITLE
What exactly does Buttons' "has-icon" do?

### DISCUSSION THREAD

**Discussion by datvm** [material-components/material-web#5043] [Q&A] [answered]
Is it like another component where it's for SSR only? Icon is shown or not currently depend on the `icon` slot being given:

- `hasIcon = false`, no `slot="icon"`:  
![image](https://github.com/material-components/material-web/assets/6388546/44d1a9b6-fccd-4007-8467-22cfa8b61f0a)

- `hasIcon = false`, there is `slot="icon"`:  
![image](https://github.com/material-components/material-web/assets/6388546/5a5809b1-5a55-4e39-9620-212010e44f7d)

- `hasIcon = true`, no `slot="icon"`:  
![image](https://github.com/material-components/material-web/assets/6388546/8498a263-3563-4352-ad97-d0a11ead5656)

- `hasIcon = true`, there is `slot="icon"`:  
![image](https://github.com/material-components/material-web/assets/6388546/c26bb9b4-1e2e-4f20-b553-713de196180d)

Is this property really needed? Can't it be deducted from `@queryAssignedElements({slot: 'icon', flatten: true})`?

**Comment by asyncliz** [comment, marked as answer]
> Is it like another component where it's for SSR only?

Yes! Due to limitations with `slotchange` logic in SSR.

These attributes are only needed to avoid a Flash Of Unstyled Content (FOUC). For the button, that means the icon popping in and changing the width when the page's javascript loads.
